src/zcash/Address.hpp

```diff
@@ -269,6 +269,17 @@ class HasShieldedRecipient {
     bool operator()(const UnifiedAddress& addr) { return true; }
 };
 
+class SelectRecipientAddress {
+public:
+    std::optional<RecipientAddress> operator()(const CKeyID& p2pkh) { return p2pkh; }
+    std::optional<RecipientAddress> operator()(const CScriptID& p2sh) { return p2sh; }
+    std::optional<RecipientAddress> operator()(const SproutPaymentAddress& addr) { return std::nullopt; }
+    std::optional<RecipientAddress> operator()(const SaplingPaymentAddress& addr) { return addr; }
+    std::optional<RecipientAddress> operator()(const UnifiedAddress& addr) {
+        return addr.GetPreferredRecipientAddress();
+    }
+};
```

Comment:
This is exactly the `RecipientForPaymentAddress` visitor I introduced in https://github.com/zcash/zcash/commit/6f1a54cd1689925a8bcae3f55eaef2a640b4d270, which you just removed in https://github.com/zcash/zcash/commit/f5d4f6fef1c3553717ca589f4bb881afd2c63898. I'll admit I'm confused as to why you went to the effort of removing `RawAddress` only to immediately re-introduce it.